src/cswinrt/strings/WinRT.cs

```diff
@@ -501,33 +547,144 @@ public unsafe ObjectReference<I> _ActivateInstance<I>()
                 MarshalInspectable<object>.DisposeAbi(instancePtr);
             }
         }
-
-#if NET
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:RequiresUnreferencedCode",
-            Justification = "No members of the generic type are dynamically accessed in this code path.")]
-#endif
-        public ObjectReference<I> _As<I>() => _IActivationFactory.As<I>();
-        public IObjectReference _As(Guid iid) => _IActivationFactory.As<WinRT.Interop.IUnknownVftbl>(iid);
     }
 
     internal sealed class ActivationFactory<T> : BaseActivationFactory
+    {
+        private static volatile ActivationFactory<T> _instance;
+
+#if NET
+        internal static ActivationFactory<T> Get()
+#else
+        internal static ActivationFactory<T> Get()
+#endif
+        {
+            var existingInstance = _instance;
+            if (existingInstance != null && (existingInstance.ContextToken == IntPtr.Zero || existingInstance.ContextToken == Context.GetContextToken()))
+            {
+                return existingInstance;
+            }
+
+            var newInstance = new ActivationFactory<T>();
+            Interlocked.CompareExchange(ref _instance, newInstance, existingInstance);
+            return newInstance;
+        }
+
+        public ActivationFactory()
+            : base(typeof(T).Namespace, typeof(T).FullName)
+        {
+        }
+    }
+
+#if NET
+    internal class BaseFactory
+    {
+        private IObjectReference _factory;
+        public IObjectReference Value { get => _factory; }
+#else
+    internal class BaseFactory<I>
     {
-        public ActivationFactory() : base(typeof(T).Namespace, typeof(T).FullName) { }
+        private ObjectReference<I> _factory;
```

Comment:
Shouldn't this also be volatile? It's being written to concurrently with volatile writes in that `cmpxgh` below.
```suggestion
        private volatile ObjectReference<I> _factory;
```